Core/src/main/java/com/plotsquared/core/util/query/PlotQuery.java

```diff
@@ -0,0 +1,228 @@
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ *
+ *     This program is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ *     You should have received a copy of the GNU General Public License
+ *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.plotsquared.core.util.query;
+
+import com.google.common.base.Preconditions;
+import com.plotsquared.core.PlotSquared;
+import com.plotsquared.core.player.PlotPlayer;
+import com.plotsquared.core.plot.Plot;
+import com.plotsquared.core.plot.PlotArea;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * This represents a plot query, and can be used to
+ * search for plots matching certain criteria.
+ * <p>
+ * The queries can be reused as no results are stored
+ * in the query itself
+ */
+public final class PlotQuery {
+
+    private PlotProvider plotProvider = new GlobalPlotProvider();
+    private final Collection<PlotFilter> filters = new LinkedList<>();
+
+    private PlotQuery() {
+    }
+
+    /**
+     * Create a new plot query instance
+     *
+     * @return New query
+     */
+    public static PlotQuery newQuery() {
+        return new PlotQuery();
+    }
+
+    /**
+     * Query for plots in a single area
+     *
+     * @param area Area
+     * @return The query instance
+     */
+    @NotNull public PlotQuery inArea(@NotNull final PlotArea area) {
+        Preconditions.checkNotNull(area, "Area may not be null");
+        this.plotProvider = new AreaLimitedPlotProvider(Collections.singletonList(area));
+        return this;
+    }
+
+    /**
+     * Query for plots in all areas in a world
+     *
+     * @param world World name
+     * @return The query instance
+     */
+    @NotNull public PlotQuery inWorld(@NotNull final String world) {
+        Preconditions.checkNotNull(world, "World may not be null");
+        this.plotProvider = new AreaLimitedPlotProvider(PlotSquared.get().getPlotAreas(world));
+        return this;
+    }
+
+    /**
+     * Query for plots in specific areas
+     *
+     * @param areas Plot areas
+     * @return The query instance
+     */
+    @NotNull public PlotQuery inAreas(@NotNull final Collection<PlotArea> areas) {
+        Preconditions.checkNotNull(areas, "Areas may not be null");
+        Preconditions.checkState(!areas.isEmpty(), "At least one area must be provided");
+        this.plotProvider = new AreaLimitedPlotProvider(Collections.unmodifiableCollection(areas));
```

Comment:
Why `Collections.umodifiableCollection(areas)`? The underlying collection may still be changed from the outside